src/udp.rs

```diff
@@ -349,142 +289,133 @@ impl Inner {
-                data
+                return Ok(true);
             };
 
-            match self.sock.send(&data) {
-                Ok(bytes_txd) => {
-                    if bytes_txd < data.len() {
-                        warn!(
-                            "Partial datagram sent. Will likely be interpreted as corrupted.
-                               Queued to be sent again."
-                        );
-                        self.current_write = Some(data);
-                    }
-                }
-                Err(error) => {
-                    if error.kind() == ErrorKind::WouldBlock
-                        || error.kind() == ErrorKind::Interrupted
-                    {
-                        self.current_write = Some(data);
-                        return Ok(false);
-                    } else {
-                        return Err(From::from(error));
-                    }
-                }
-            }
+            let res = self.sock.send(&data);
+            handle_send_res(res, &mut self.current_write, data.len(), data)?;
         }
     }
 
-    fn write_to<T: Serialize>(&mut self, msg: Option<(T, SocketAddr, Priority)>) -> ::Res<bool> {
-        let expired_keys: Vec<u8> = self
-            .write_queue_2
-            .iter()
-            .skip_while(|&(&priority, queue)| {
-                priority < self.conf.msg_drop_priority || // Don't drop high-priority messages.
-                queue.front().map_or(true, |&(ref timestamp, _, _)| {
-                    timestamp.elapsed().as_secs() <= self.conf.max_msg_age_secs
-                })
-            }).map(|(&priority, _)| priority)
-            .collect();
-        let dropped_msgs: usize = expired_keys
-            .iter()
-            .filter_map(|priority| self.write_queue_2.remove(priority))
-            .map(|queue| queue.len())
-            .sum();
-        if dropped_msgs > 0 {
-            trace!(
-                "Insufficient bandwidth. Dropping {} messages with priority >= {}.",
-                dropped_msgs,
-                expired_keys[0]
-            );
-        }
+    /// Returns `Ok(false)`, if write to the underlying stream would block.
+    fn flush_write_to_until_would_block(&mut self) -> ::Res<bool> {
+        loop {
+            let (data, peer) = if let Some(data_peer) = self
+                .current_write_2
+                .take()
+                .or_else(|| self.out_queue2.next_msg())
+            {
+                data_peer
+            } else {
+                return Ok(true);
+            };
 
-        if let Some((msg, peer, priority)) = msg {
-            let entry = self
-                .write_queue_2
-                .entry(priority)
-                .or_insert_with(|| VecDeque::with_capacity(10));
-            entry.push_back((Instant::now(), serialise(&msg)?, peer));
+            let res = self.sock.send_to(&data, &peer);
+            handle_send_res(res, &mut self.current_write_2, data.len(), (data, peer))?;
```

Comment:
Why is it not breaking out of the loop for `Ok(false)` ? At that point you will have triggered a wouldblock so you shouldn't really be attempting to do anything with the socket until the next edge-trigger for write readiness.

Do check though if that would require any other change.